docs/Developers/Design/CRD_design.md

```diff
@@ -0,0 +1,438 @@
+via ithe cli `gluster vol set`) that do not take a volume parameter.
+
+The `glusterCA` field holds a reference to a Kubernetes Secret containing the
+certificate authority `.key` and `.pem` files from which both client and server
+TLS keys can be generated. These will be used to automatically configure data
+encryption between the CSI driver and the Gluster bricks.
+
+The `replication` set of parameters define the geo-replication configuration
+for this cluster, optionally as both a source and target. If this cluster is to
+be used as a target for replication, the `replication.credentials` field must
+be supplied. This is a reference to a Secret that contains the inbound ssh user
+& key information. If this cluster is used as a source, the replication targets
+are presented as a list in `replication.targets`, providing a `name` for each
+remote cluster, the address(es) via `address`, and the ssh credentials via the
+`credentials` field.
+
+## Zone CR
+
+The `Zone` CR is used to define a failure domain. There would be at least one
+of these associated with a Gluster cluster. The purpose is to define how the
+operator can create new Gluster nodes within the failure domain and the rules
+for doing so. Additionally, it provides a name for the domain so that it can be
+referenced in the StorageClass to control volume placement and spreading.
+
+```yaml
+apiVersion: "operator.gluster.org/v1alpha1"
+kind: Zone
+metadata:
+  # Name for this zone
+  name: az1
+  # Zones are cluster-specific
+  cluster: my-cluster
+  # CRD is namespaced
+  namespace: gcs
+spec:
+  management:
+    # How the zone is managed
+    mode: automatic # (automatic | external | manual)
+    automatic:  # (iff mode==automatic)
+      # Minimum Gluster pods
+      minNodes: 1
+      # Maximum Gluster pods
+      maxNodes: 99
+      # Maximum raw storage capacity (South) in this zone
+      maxStorage: 100Ti
+      # Min unallocated capacity in the zone
+      freeStorageMin: 1Ti
+      # Max unallocated capacity in the zone
+      freeStorageMax: 3Ti
+    external:  # (iff mode==external)
+      nodes:
+        - 1.2.3.4
+        - some.node.net
+      # Credentials for connecting to external nodes (ssh user & key)
+      credentials: SecretRef
+    manual:  # (iff mode==manual)
+      nodes: 2
+  nodeAffinity: # https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+        - matchExpressions:
+          - key: failure-domain.beta.kubernetes.io/zone
+            operator: In
+            values:
+              - us-east-2a
+  storage:
+    # Whether to use block-mode PVs or raw /dev devices
+    # mode==device will cause /dev to be added to the PodSpec as a
+    # hostPath mount
+    mode: device # (device | volume)
+    # In device mode, the pod would use the entire device(s) as a single VG
+    device:  # (iff mode==device)
```

Comment:
Not sure, but I thought they belongs to `Node CR` ? Is it guaranteed to be same for all nodes in the zone?

Other than this, looks fine on the top level!

Reply:
That's a tough one. Here's my reasoning...
The Zone provides 2 functions:
- A grouping of nodes in a single fault domain
- A template for how to deploy nodes
It's the 2nd function that resulted in the whole `storage` subtree. The main option that is expected to be used is `mode: volume` to get block PVs. In this case, a standard device template for the whole Zone is fairly easy to see. `mode: device` is really an escape hatch for clusters that (1) don't yet support block PV or (2) are using Heketi & the old stack. I expect even bare metal to use block PV along w/ dynamic local provisioning once that is fully merged.

The other difficulty of using the Node object is that (as currently planned) it will always be created by the operator. In that case, there's no way to set the `device` list when the object is first created. Either we'd have to find a way to change the semantics or an admin would need to annotate each Node object after the operator creates it :-1: .

I do agree with you that having a fixed device list for a whole zone isn't desirable. The question is how common it is for the device list to vary by host. I would expect most environments to be fairly symmetric.

But here's an addl wrinkle that I hadn't considered...
We don't really want to use `sd*` designations since they are not stable. Instead, we'd need something in the `/dev/disk/by-*` tree, and these will vary by host. Perhaps the `sd*` can be used as a "template" in keeping w/ the Zone notion, and the actual IDs get used internally (and potentially annotated onto the Node object).

Reply:
While targeting a symmetric and automated cluster model is definitely the way to go, we can't ignore that on-prem scenarios are especially likely to not be symmetric. I've run across a number of them when dealing with customer cases. We also can't deny the reality that local block PVs may not reach beta for a good while after we want the operator to be released. I say the way to do this is to leverage (or mimic?) some of the functionality of the local storage provisioner such that we provide a regular expression of devices that are valid for storage consumption.

> We don't really want to use sd* designations since they are not stable. Instead, we'd need something in the /dev/disk/by-* tree, and these will vary by host. Perhaps the sd* can be used as a "template" in keeping w/ the Zone notion, and the actual IDs get used internally (and potentially annotated onto the Node object).

Could we have the operator detect a `/dev/disk` name for a given `/dev/sd*` name under the covers? We could update the status of the specific Node CR to indicate which device names are actually in use.